experimental/uefi/baremetal/src/main.rs

```diff
@@ -18,34 +18,38 @@
 #![no_main]
 #![feature(lang_items)]
 #![feature(alloc_error_handler)]
-#![feature(bench_black_box)]
+#![feature(core_ffi_c)]
 #![feature(custom_test_frameworks)]
 // As we're in a `no_std` environment, testing requires special handling. This
 // approach was inspired by https://os.phil-opp.com/testing/.
 #![test_runner(crate::test_runner)]
 #![reexport_test_harness_main = "test_main"]
 
 use core::panic::PanicInfo;
-use rust_hypervisor_firmware_boot::pvh;
+
+mod asm;
 mod hvm_start_info;
+mod multiboot;
 
 #[no_mangle]
 #[cfg(test)]
 pub extern "C" fn rust64_start(_rdi: &hvm_start_info::hvm_start_info) -> ! {
-    // Ensure that `PVH_NOTE` is not optimized away, otherwise we won't be able to boot our
-    // kernel when testing.
-    core::hint::black_box(&pvh::PVH_NOTE);
     test_main();
     kernel::i8042::shutdown();
 }
 
 #[no_mangle]
 #[cfg(not(test))]
-pub extern "C" fn rust64_start(rdi: &hvm_start_info::hvm_start_info) -> ! {
-    // Ensure that `PVH_NOTE` is not optimized away, otherwise we won't be able to boot our
-    // kernel when testing.
-    core::hint::black_box(&pvh::PVH_NOTE);
-    kernel::start_kernel(rdi);
+pub extern "C" fn rust64_start(start_info: usize, magic: u64) -> ! {
+    if magic == 0x2BADB002 {
```

Comment:
Is there a reference where this is defined (like a spec)? If so, perhaps add a link in a comment.

Reply:
Done. I feature-gated the whole multiboot main entry point as well, as that let me get rid of the ugly unsafe block, below.